api/service/stagedsync/stage_blockhashes.go

```diff
@@ -0,0 +1,600 @@
+func (bh *StageBlockHashes) loadBlockHashesFromCache(s *StageState, startHeight uint64, targetHeight uint64, tx kv.RwTx) (p uint64, h common.Hash, err error) {
+
+	p = startHeight
+	var lastHash []byte
+	h.SetBytes(lastHash[:])
+	useInternalTx := tx == nil
+	if useInternalTx {
+		tx, err = bh.configs.db.BeginRw(bh.configs.ctx)
+		if err != nil {
+			return p, h, err
+		}
+		defer tx.Rollback()
+	}
+
+	if errV := bh.configs.cachedb.View(context.Background(), func(rtx kv.Tx) error {
+		// load block hashes from cache db and copy them to main sync db
+		for ok := true; ok; ok = p < targetHeight {
+			key := strconv.FormatUint(p+1, 10)
+			lastHash, err = rtx.GetOne(BlockHashesBucket, []byte(key))
+			if err != nil {
+				utils.Logger().Warn().
+					Err(err).
+					Str("block height", key).
+					Msg("[STAGED_SYNC] retrieve block hash from cache failed")
+				return err
+			}
+			if len(lastHash[:]) == 0 {
+				return nil
+			}
+			bucketName := GetBucketName(BlockHashesBucket, s.state.isBeacon)
+			if err = tx.Put(bucketName, []byte(key), lastHash); err != nil {
+				return err
+			}
+			p++
+		}
+		// load extra block hashes from cache db and copy them to bg db to be downloaded in background by block stage
+		pExtraHashes := p
+		for ok := true; ok; ok = pExtraHashes < p+s.state.MaxBackgroundBlocks {
+			key := strconv.FormatUint(pExtraHashes+1, 10)
+			newHash, err := rtx.GetOne(BlockHashesBucket, []byte(key))
+			if err != nil {
+				utils.Logger().Warn().
+					Err(err).
+					Str("block height", key).
+					Msg("[STAGED_SYNC] retrieve extra block hashes for background process failed")
+				break
+			}
+			if len(newHash[:]) == 0 {
+				return nil
+			}
+			bucketName := GetBucketName(ExtraBlockHashesBucket, s.state.isBeacon)
+			if err = tx.Put(bucketName, []byte(key), newHash); err != nil {
+				break
+			}
+			pExtraHashes++
+		}
+		return nil
+	}); errV != nil {
+		return startHeight, h, errV
+	}
+
+	// set last hash
+	h.SetBytes(lastHash[:])
+
+	// save progress
+	if err = s.Update(tx, p); err != nil {
+		utils.Logger().Info().
+			Msgf("[STAGED_SYNC] saving retrieved cached progress for block hashes stage failed: %v", err)
+		return startHeight, h, err
+	}
+
+	// update the progress
+	if useInternalTx {
+		if err := tx.Commit(); err != nil {
+			return startHeight, h, err
+		}
+	}
+	return p, h, nil
+}
+
+func (bh *StageBlockHashes) Unwind(firstCycle bool, u *UnwindState, s *StageState, tx kv.RwTx) (err error) {
+	useInternalTx := tx == nil
+	if useInternalTx {
+		tx, err = bh.configs.db.BeginRw(bh.configs.ctx)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback()
+	}
+
+	if err = u.Done(tx); err != nil {
+		return fmt.Errorf(" reset: %w", err)
+	}
+
+	if useInternalTx {
+		if err = tx.Commit(); err != nil {
+			return ErrCommitTransactionFail
+		}
+	}
+	return nil
+}
+
+func (bh *StageBlockHashes) Prune(firstCycle bool, p *PruneState, tx kv.RwTx) (err error) {
```

Comment:
I know unwind is to revert the blocks. What does Prune do?

Reply:
Please add comment

Reply:
To prune whatever temporary is used by staged. There is a PruneFunc for each stage which is the execution function for the stage to prune old data.Basically it's like a cleanup for the stage.

Reply:
Please rename to cleanup. Prune has a special meaning of pruning part of the data from a whole. This sounds more like a cleanup of junk/useless data.

Reply:
Renamed!